planning/motion_velocity_planner/autoware_motion_velocity_obstacle_stop_module/src/path_length_buffer.hpp

```diff
@@ -0,0 +1,177 @@
+    bool active, const StopObstacle & determined_stop_obstacle,
+    double determined_desired_stop_margin)
+  : stop_distance(stop_distance),
+    stop_point(point),
+    start_time(time),
+    is_active(active),
+    determined_stop_obstacle(determined_stop_obstacle),
+    determined_desired_stop_margin(determined_desired_stop_margin)
+  {
+  }
+
+  bool should_activate(const rclcpp::Time & current_time, double min_on_duration) const
+  {
+    return !is_active && (current_time - start_time).seconds() > min_on_duration;
+  }
+};
+
+class PathLengthBuffer
+{
+public:
+  PathLengthBuffer() = default;
+  PathLengthBuffer(
+    double update_distance_threshold, double min_off_duration, double min_on_duration)
+  : min_off_duration_(min_off_duration),
+    min_on_duration_(min_on_duration),
+    update_distance_th_(update_distance_threshold)
+  {
+  }
+
+  std::optional<BufferedStopDistanceItem> get_nearest_active_item() const
+  {
+    if (buffer_.empty()) {
+      return {};
+    }
+
+    auto nearest_item = std::min_element(
+      buffer_.begin(), buffer_.end(),
+      [](const BufferedStopDistanceItem & a, const BufferedStopDistanceItem & b) {
+        if (!a.is_active) return false;
+        if (!b.is_active) return true;
+        return a.stop_distance < b.stop_distance;
+      });
+
+    if (!nearest_item->is_active) {
+      return std::nullopt;
+    }
+
+    return *nearest_item;
+  }
+
+  void update_buffer(
+    const std::optional<geometry_msgs::msg::Point> & stop_point,
+    std::function<double(const geometry_msgs::msg::Point &)> calc_stop_distance,
+    const rclcpp::Time & current_time, const StopObstacle & determined_stop_obstacle,
+    double determined_desired_stop_margin)
+  {
+    if (!stop_point) {
+      return;
+    }
+
+    const double stop_distance = calc_stop_distance(*stop_point);
+
+    // Update the state of buffered-items
+    for (auto & buffered_item : buffer_) {
+      // Update the stop distance
+      buffered_item.stop_distance = calc_stop_distance(buffered_item.stop_point);
+
+      // Handle timing-based activation
+      const bool should_activate = buffered_item.should_activate(current_time, min_on_duration_);
+      if (!should_activate) {
+        continue;
+      }
+
+      buffered_item.is_active = true;
+      buffered_item.start_time = current_time;
+    }
+
+    // Remove items that are invalid
+    buffer_.erase(
+      std::remove_if(
+        buffer_.begin(), buffer_.end(),
+        [&](const BufferedStopDistanceItem & buffered_item) {
+          const auto duration = (current_time - buffered_item.start_time).seconds();
+
+          const double rel_dist = std::abs(buffered_item.stop_distance - stop_distance);
+
+          const bool is_remove = (buffered_item.is_active && (duration > min_off_duration_)) ||
+                                 (!buffered_item.is_active && rel_dist > update_distance_th_);
+
+          return is_remove;
```

Comment:
Please always consider RVO.

```suggestion
          return (buffered_item.is_active && (duration > min_off_duration_)) || (!buffered_item.is_active && rel_dist > update_distance_th_);
```

Reply:
Done